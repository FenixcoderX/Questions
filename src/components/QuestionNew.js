import { connect } from 'react-redux';
import { useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { handleСreateQuestion } from '../actions/questions';
// import { handleSaveQuestionAnswer } from '../actions/questions';

const QuestionNew = ({ dispatch }) => {
  const [firstOption, setFirstOption] = useState('');
  const [secondOption, setSecondOption] = useState('');
  const navigate = useNavigate();

  const handleSubmit = (e) => {
    e.preventDefault();
    dispatch(handleСreateQuestion(firstOption, secondOption));
    setFirstOption('');
    setSecondOption('');
    navigate(`/`);
  };

  const handleChangeFirstOption = (e) => {
    const firstOption = e.target.value;
    setFirstOption(firstOption);
  };
  const handleChangeSecondOption = (e) => {
    const secondOption = e.target.value;
    setSecondOption(secondOption);
  };

  return (
    <div className="question-new-container">
      <h3 className="question-new-header">Create new question</h3>
      <div className="question-new-header">Help Me Choose</div>
      <form className="textarea-form mb-3" onSubmit={handleSubmit}>
        <label className="form-label">First option</label>
        <textarea
          itemID=""
          data-testid="first-option-input"
          placeholder="maximum 100 characters"
          value={firstOption}
          onChange={handleChangeFirstOption}
          className="form-control"
          rows="3"
          maxLength={100}
        ></textarea>
        <label className="form-label">Second option</label>
        <textarea
          data-testid="second-option-input"
          placeholder="maximum 100 characters"
          value={secondOption}
          onChange={handleChangeSecondOption}
          className="form-control"
          rows="3"
          maxLength={100}
        ></textarea>
        <button
          data-testid="create-button"
          className="btn btn-dark text-nowrap"
          type="submit"
          disabled={firstOption === '' || secondOption === ''}
        >
          Send
        </button>
      </form>
    </div>
  );
};

const mapStateToProps = () => {
  return {};
};

export default connect(mapStateToProps)(QuestionNew);
