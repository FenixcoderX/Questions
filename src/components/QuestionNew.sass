.question-new-container
  display: flex
  flex-direction: column
  flex-wrap: wrap
  align-items: center
  text-align: center
  padding: 10px

.question-new-header
  white-space: nowrap
  margin-bottom: 15px
  font-weight: 600

.question-new-subheader
  margin-bottom: 15px

.question-new-textarea-form
  display: flex
  flex-direction: column
  border-radius: 13px
  margin-right: auto
  margin-left: auto
  padding: 10px
  width: 70%
  min-width: 280px
  max-width: 500px
  background-color: white
  align-items: center
  text-align: center

  textarea
    margin: 10px
    text-align: center
    align-content: center

    &::placeholder
      font-weight: 300

  button
    border-radius: 35px