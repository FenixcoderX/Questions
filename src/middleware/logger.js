/**
 * Middleware function that logs the action and the new state to the console.
 */
const logger = (store) => (next) => (action) => {
  console.group(action.type);
  console.log('The action: ', action);
  const returnValue = next(action);                   // Calls dispatch, because it should be last middleware in the chain
  console.log('The new state: ', store.getState());
  console.groupEnd();
  return returnValue;                                  // Return the result that we have after invoking next
};

export default logger;
