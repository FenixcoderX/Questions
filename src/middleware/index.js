import thunk from "redux-thunk";
import logger from "./logger";
import { applyMiddleware } from "redux";

// Middleware will be applied in point between dispatching an action, and the moment it reaches the reducer. In the order it is listed
export default applyMiddleware(thunk, logger);
