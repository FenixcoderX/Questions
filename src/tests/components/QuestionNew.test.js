import React from 'react';
import QuestionNew from '../../components/QuestionNew';
import { BrowserRouter as Router } from 'react-router-dom';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
import reducer from '../../reducers';
import middleware from '../../middleware';
import { render, fireEvent } from '@testing-library/react';

const store = createStore(reducer, middleware);

describe('QuestionNew component', () => {
  let questionNewComponent;
  let firstOptionInput;
  let secondOptionInput;
  let createButton;

  beforeEach(() => {
    questionNewComponent = render(
      <Provider store={store}>
        <Router>
          <QuestionNew />
        </Router>
      </Provider>
    );

    firstOptionInput = questionNewComponent.getByTestId('first-option-input');
    secondOptionInput = questionNewComponent.getByTestId('second-option-input');
    createButton = questionNewComponent.getByTestId('create-button');
  });

  it('should show First option input, Second option input, disabled Create button in the document', () => {
    expect(firstOptionInput).toBeInTheDocument();
    expect(secondOptionInput).toBeInTheDocument();
    expect(createButton).toBeInTheDocument();
    expect(createButton).toBeDisabled();
  });

  it('should match snapshot', () => {
    expect(questionNewComponent).toMatchSnapshot();
  });

  it('should enable the Create button if input fields are filled', async () => {
    fireEvent.change(firstOptionInput, {
      target: { value: 'first' },
    });
    fireEvent.change(secondOptionInput, {
      target: { value: 'second' },
    });

    expect(createButton).toBeEnabled();
  });
  it('should disable the Create button if not all input fields are filled', () => {
    fireEvent.change(firstOptionInput, {
      target: { value: 'first' },
    });
    fireEvent.change(secondOptionInput, { target: { value: '' } });

    expect(createButton).toBeDisabled();
  });
});
