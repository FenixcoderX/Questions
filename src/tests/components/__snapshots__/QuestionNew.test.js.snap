// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionNew component should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="question-new-container"
      >
        <h3
          class="question-new-header"
        >
          Create new question
        </h3>
        <div
          class="question-new-header"
        >
          Help Me Choose
        </div>
        <form
          class="textarea-form mb-3"
        >
          <label
            class="form-label"
          >
            First option
          </label>
          <textarea
            class="form-control"
            data-testid="first-option-input"
            itemid=""
            maxlength="100"
            placeholder="maximum 100 characters"
            rows="3"
          />
          <label
            class="form-label"
          >
            Second option
          </label>
          <textarea
            class="form-control"
            data-testid="second-option-input"
            maxlength="100"
            placeholder="maximum 100 characters"
            rows="3"
          />
          <button
            class="btn btn-dark text-nowrap"
            data-testid="create-button"
            disabled=""
            type="submit"
          >
            Send
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="question-new-container"
    >
      <h3
        class="question-new-header"
      >
        Create new question
      </h3>
      <div
        class="question-new-header"
      >
        Help Me Choose
      </div>
      <form
        class="textarea-form mb-3"
      >
        <label
          class="form-label"
        >
          First option
        </label>
        <textarea
          class="form-control"
          data-testid="first-option-input"
          itemid=""
          maxlength="100"
          placeholder="maximum 100 characters"
          rows="3"
        />
        <label
          class="form-label"
        >
          Second option
        </label>
        <textarea
          class="form-control"
          data-testid="second-option-input"
          maxlength="100"
          placeholder="maximum 100 characters"
          rows="3"
        />
        <button
          class="btn btn-dark text-nowrap"
          data-testid="create-button"
          disabled=""
          type="submit"
        >
          Send
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
